gr-zeromq/lib/base_impl.cc

```diff
@@ -85,7 +85,12 @@ base_impl::base_impl(int type,
     }
 }
 
-base_impl::~base_impl() {}
+base_impl::~base_impl()
+{
+    d_context.shutdown();
+    d_socket.close();
+    d_context.close();
```

Comment:
The `zmq::socket_t` destructor [already calls `close()`](https://github.com/zeromq/cppzmq/blob/160ac8ec61298508093933a9f53bfedfb6ba0425/zmq.hpp#L2225), and likewise the `zmq::context_t` destructor [already calls `close()`](https://github.com/zeromq/cppzmq/blob/160ac8ec61298508093933a9f53bfedfb6ba0425/zmq.hpp#L832). A double `close()` isn't harmful because on the first `close()` an internal pointer is set to `nullptr`, so the second `close()` does nothing. But I would say that implicitly calling `close()` with the destructors is better RAII style.

The call to `shutdown()` is interesting. It basically calls [`zmq_ctx_shutdown()`](http://api.zeromq.org/master:zmq-ctx-shutdown), which says "Context shutdown will cause any blocking operations currently in progress on sockets open within context to return immediately with an error code of ETERM." This seems to be something we'd like, because if we have a ZMQ block blocked waiting for data, we'd want to be able to terminate the flowgraph normally (I currently have some use cases in which the flowgraph needs to be SIGKILL'ed because of this). However, I wonder if the `shutdown()` here really achieves what is intended. If a call to `zmq::poll()` (or whatever other function) is blocked in the `work()` function, can the destructor actually start running?

Reply:
That's a good point. It was the shutdown I thought was the most interesting. I wonder if the shutdown belongs in `stop()` instead of in the destructor right before the `join()`. @SaikWolf any thoughts on this? Did you find that the shutdown had an effect after the flowgraph was already stopped?

Reply:
My reason for it in the destructor over the stop:
 - There might be cases where the flowgraph has been stopped, but any processing messages might still be desired.
 - If the object is being destroyed, it's clear that anything in process can be dropped.

Reply:
The reason for the explicit close calls is that the context/socket need to be closed in that order. Sometimes I've seen them closed in the wrong order causing a hang. This just makes sure everything wraps up as ZMQ likes.

Reply:
Given that the class members are declared in the order
```c++
    zmq::context_t d_context;
    zmq::socket_t d_socket;
```
(which makes sense, because to construct `d_socket` we need `d_context), the destructors will run in the reverse order: first for the socket, then for the context, which is the correct order for the `close()`'s. What are the cases in which you observed the wrong order?

Reply:
It's usually been in python operations. I could be remembering wrong, and this is overkill.